streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java

```diff
@@ -74,20 +209,240 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareStandbyTas
+
+    /**
+     * This function returns whether the current application state has the required rack information
+     * to make assignment decisions with.
+     * This includes ensuring that every client has a known rack id, and that every topic partition for
+     * every logical task that needs to be assigned also has known rack ids.
+     * If a logical task has no source topic partitions, it will return false.
+     * If standby tasks are configured, but a logical task has no changelog topic partitions, it will return false.
      *
-     * @return a new map containing an assignment that replicates exactly the previous assignment reported
-     *         in the applicationState
+     * @return whether rack-aware assignment decisions can be made for this application.
      */
-    public static Map<ProcessId, KafkaStreamsAssignment> identityAssignment(
-        final ApplicationState applicationState
-    ) {
-        throw new UnsupportedOperationException("Not Implemented.");
+    public static boolean hasValidRackInformation(final ApplicationState applicationState) {
+        for (final KafkaStreamsState state : applicationState.kafkaStreamsStates(false).values()) {
+            if (!hasValidRackInformation(state)) {
+                return false;
+            }
+        }
+
+        final boolean hasStandby = applicationState.assignmentConfigs().numStandbyReplicas() >= 1;
+        for (final TaskInfo task : applicationState.allTasks()) {
+            if (!hasValidRackInformation(task, hasStandby)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private static boolean hasValidRackInformation(final KafkaStreamsState state) {
+        if (!state.rackId().isPresent()) {
+            LOG.error("Client " + state.processId() + " doesn't have rack configured.");
+            return false;
+        }
+        return true;
+    }
+
+    private static boolean hasValidRackInformation(final TaskInfo task, final boolean hasStandby) {
+        final Map<TopicPartition, Set<String>> racksByTopicPartition = task.partitionToRackIds();
+        for (final TopicPartition topicPartition : task.sourceTopicPartitions()) {
+            final Set<String> racks = racksByTopicPartition.getOrDefault(topicPartition, null);
+            if (racks == null || racks.isEmpty()) {
+                LOG.error("Topic partition {} for task {} does not have racks configured.", topicPartition, task.id());
+                return false;
+            }
+        }
+
+        for (final TopicPartition topicPartition : task.changelogTopicPartitions()) {
+            final Set<String> racks = racksByTopicPartition.getOrDefault(topicPartition, null);
+            if (racks == null || racks.isEmpty()) {
+                LOG.error("Topic partition {} for task {} does not have racks configured.", topicPartition, task.id());
+                return false;
+            }
+        }
+
+        if (task.sourceTopicPartitions().isEmpty()) {
```

Comment:
It really doesn't make sense to include this check here, whether a task has source partitions has nothing to do with whether the rack info is valid. Also, it shouldn't be up to the user/assignor to verify the inputs to the assignor and check conditions such as "every task has at least one source topic partition".